src/pd/async/util.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+use std::sync::RwLock;
+use std::time::Instant;
+use std::time::Duration;
+use std::thread;
+
+use futures::Future;
+use futures::future::ok;
+use futures::future::{self, loop_fn, Loop};
+
+use kvproto::pdpb::GetMembersResponse;
+use kvproto::pdpb_grpc::PDAsyncClient;
+
+use util::HandyRwLock;
+
+use super::super::PdFuture;
+use super::super::Result;
+use super::client::try_connect_leader;
+
+pub struct Inner {
+    pub client: Arc<PDAsyncClient>,
+    pub members: GetMembersResponse,
+}
+
+/// A leader client doing requests asynchronous.
+pub struct LeaderClient {
+    pub inner: Arc<RwLock<Inner>>,
+}
+
+impl LeaderClient {
+    pub fn new(client: PDAsyncClient, members: GetMembersResponse) -> LeaderClient {
+        LeaderClient {
+            inner: Arc::new(RwLock::new(Inner {
+                client: Arc::new(client),
+                members: members,
+            })),
+        }
+    }
+
+    pub fn request<Req, Resp, F>(&self, req: Req, f: F, retry: usize) -> Request<Req, Resp, F>
+        where Req: Clone + 'static,
+              F: FnMut(&PDAsyncClient, Req) -> PdFuture<Resp> + Send + 'static
+    {
+        Request {
+            reconnect_count: retry,
+            request_sent: 0,
+            inner: self.inner.clone(),
+            req: req,
+            resp: None,
+            client: None,
+            func: f,
+        }
+    }
+
+    pub fn get_client(&self) -> Arc<PDAsyncClient> {
+        self.inner.rl().client.clone()
+    }
+
+    pub fn get_members(&self) -> GetMembersResponse {
```

Comment:
`GetMembersResponse` is locked in `RwLock`, so you suggest returning a `RwLockReadGuard<T>` ?